utils/test_utils.py

```diff
@@ -0,0 +1,259 @@
+            'a9.gcode',
+            '10.gcode',
+            '11.gcode']
+    )
+
+def test_generate_new_filename_abc_7():
+    simulate_printing_12_times_and_validate_result(
+        filename='abc.gcode',
+        max_filename=7,
+        existing_sd_files=[],
+        expected_sd_files=[
+            'a.gcode',
+            '1.gcode',
+            '2.gcode',
+            '3.gcode',
+            '4.gcode',
+            '5.gcode',
+            '6.gcode',
+            '7.gcode',
+            '8.gcode',
+            '9.gcode',
+            '10.gcode',
+            '11.gcode']
+    )
+
+def test_generate_new_filename_abc_6():
+    simulate_printing_12_times_and_validate_result(
+        filename='abc.gcode',
+        max_filename=6,
+        existing_sd_files=[],
+        expected_sd_files=[
+            '.gcode',
+            '1.gcode',
+            '2.gcode',
+            '3.gcode',
+            '4.gcode',
+            '5.gcode',
+            '6.gcode',
+            '7.gcode',
+            '8.gcode',
+            '9.gcode',
+            '10.gcode',
+            '11.gcode']
+    )
+
+def test_generate_new_filename_abc_5():
+    simulate_printing_12_times_and_validate_result(
+        filename='abc.gcode',
+        max_filename=5,
+        existing_sd_files=[],
+        expected_sd_files=[
+            '.gcode',
+            '1.gcode',
+            '2.gcode',
+            '3.gcode',
+            '4.gcode',
+            '5.gcode',
+            '6.gcode',
+            '7.gcode',
+            '8.gcode',
+            '9.gcode',
+            '10.gcode',
+            '11.gcode']
+    )
+
+def test_generate_new_filename_abc_0():
+    simulate_printing_12_times_and_validate_result(
+        filename='abc.gcode',
+        max_filename=0,
+        existing_sd_files=[],
+        expected_sd_files=[
+            '.gcode',
+            '1.gcode',
+            '2.gcode',
+            '3.gcode',
+            '4.gcode',
+            '5.gcode',
+            '6.gcode',
+            '7.gcode',
+            '8.gcode',
+            '9.gcode',
+            '10.gcode',
+            '11.gcode']
+    )
+
+def test_generate_new_filename_abc1_30_existing_abc1():
+    simulate_printing_12_times_and_validate_result(
+        filename='abc1.gcode',
+        max_filename=30,
+        existing_sd_files=['abc1.gcode'],
+        expected_sd_files=[
+            'abc1.gcode',
+            'abc11.gcode',
+            'abc12.gcode',
+            'abc13.gcode',
+            'abc14.gcode',
+            'abc15.gcode',
+            'abc16.gcode',
+            'abc17.gcode',
+            'abc18.gcode',
+            'abc19.gcode',
+            'abc110.gcode',
+            'abc111.gcode',
+            'abc112.gcode']
+    )
+
+def test_generate_new_filename_abc_30_existing_abc5():
+    simulate_printing_12_times_and_validate_result(
+        filename='abc.gcode',
+        max_filename=30,
+        existing_sd_files=['abc5.gcode'],
+        expected_sd_files=[
+            'abc5.gcode',
```

Comment:
It looks like we can't name it as **abc5.gcode** because it already exists

Reply:
this is the purpose of the PR, if such file already exists then just generate new name :)

Reply:
yeah, silly me. I missed that the **existing_sd_files** it's a function parameter, not a local variable. Resolved